server/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -0,0 +1,120 @@
+package org.http4s.server.middleware
+
+import org.http4s.{Headers, Http, Response, Status}
+import cats.data.Kleisli
+import cats.effect.{Clock, Sync}
+import cats.effect.concurrent.Ref
+import scala.concurrent.duration.FiniteDuration
+import cats.implicits._
+import java.util.concurrent.TimeUnit.NANOSECONDS
+import org.http4s.headers.`Retry-After`
+
+/**
+  * Transform a service to reject any calls the go over a given rate.
+  */
+object Throttle {
+
+  sealed abstract class TokenAvailability extends Product with Serializable
+  case class TokenAvailable() extends TokenAvailability
+  case class TokenUnavailable(retryAfter: Option[FiniteDuration]) extends TokenAvailability
+
+  /**
+    * A token bucket for use with the [[Throttle]] middleware.  Consumers can take tokens which will be refilled over time.
+    * Implementations are required to provide their own refill mechanism.
+    *
+    * Possible implementations include a remote TokenBucket service to coordinate between different application instances.
+    */
+  trait TokenBucket[F[_]] {
+    def takeToken: F[TokenAvailability]
+  }
+
+  object TokenBucket {
+
+    /**
+      * Creates an in-memory [[TokenBucket]].
+      *
+      * @param capacity the number of tokens the bucket can hold and starts with.
+      * @param refillEvery the frequency with which to add another token if there is capacity spare.
+      * @return A task to create the [[TokenBucket]].
+      */
+    def local[F[_]](capacity: Int, refillEvery: FiniteDuration)(
+      implicit F: Sync[F],
+      clock: Clock[F]): F[TokenBucket[F]] = {
+
+      def getTime = clock.monotonic(NANOSECONDS)
+      val bucket = getTime.flatMap(time => Ref[F].of((capacity.toDouble, time)))
+
+      bucket.map { counter =>
+        new TokenBucket[F] {
+          override def takeToken: F[TokenAvailability] = {
+            val attemptUpdate = counter.access.flatMap { case ((previousTokens, previousTime), setter) =>
+              getTime.flatMap( currentTime => {
+                val timeDifference = currentTime - previousTime
+                val tokensToAdd = timeDifference.toDouble / refillEvery.toNanos.toDouble
+                val newTokenTotal = Math.min(previousTokens + tokensToAdd, capacity.toDouble)
+
+                val attemptSet: F[Option[TokenAvailability]] = if (newTokenTotal >= 1) {
+                  setter((newTokenTotal - 1, currentTime)).map(success => if (success) Some(TokenAvailable()) else None)
+                } else {
+                  val timeToFull = (refillEvery.toNanos * capacity) - timeDifference
+                  val successResponse = TokenUnavailable(Some(FiniteDuration(timeToFull, NANOSECONDS)))
+                  setter((newTokenTotal, currentTime)).map(success => if (success) Some(successResponse) else None)
```

Comment:
`TokenUnavailable(Some(FiniteDuration(timeToFull, NANOSECONDS)))`

apologies for the bikeshedding (feel free to ignore)

`TokenUnavailable(timeToFull.nanos.some)`

Reply:
Bikeshedding appreciated, I had the feeling there would be a more concise way to do this...